src/final_chain/final_chain.cpp

```diff
@@ -75,43 +76,40 @@ class FinalChainImpl final : public FinalChain {
     }
   }
 
-  future<shared_ptr<FinalizationResult const>> finalize(NewBlock new_blk,
+  future<shared_ptr<FinalizationResult const>> finalize(NewBlock new_blk, uint64_t period,
                                                         finalize_precommit_ext precommit_ext = {}) override {
     auto p = make_shared<promise<shared_ptr<FinalizationResult const>>>();
-    executor_([this, new_blk = move(new_blk), precommit_ext = move(precommit_ext), p]() mutable {
-      p->set_value(finalize_(move(new_blk), precommit_ext));
+    executor_([this, new_blk = move(new_blk), period, precommit_ext = move(precommit_ext), p]() mutable {
+      p->set_value(finalize_(move(new_blk), period, precommit_ext));
     });
     return p->get_future();
   }
 
-  shared_ptr<FinalizationResult> finalize_(NewBlock new_blk, finalize_precommit_ext const& precommit_ext) {
+  shared_ptr<FinalizationResult> finalize_(NewBlock new_blk, uint64_t period,
+                                           finalize_precommit_ext const& precommit_ext) {
     auto batch = db_->createWriteBatch();
     Transactions to_execute;
     {
       // This artificial scope will make sure we clean up the big chunk of memory allocated for this batch-processing
       // stuff as soon as possible
-      DB::MultiGetQuery db_query(db_, 2 * transaction_count_hint_ + new_blk.dag_blk_hashes.size());
-      auto dag_blks_raw = db_query.append(DB::Columns::dag_blocks, new_blk.dag_blk_hashes, false).execute();
-      unordered_set<h256> unique_trxs;
-      unique_trxs.reserve(transaction_count_hint_);
-      for (auto const& dag_blk_raw : dag_blks_raw) {
-        for (auto const& trx_h : DagBlock::extract_transactions_from_rlp(RLP(dag_blk_raw))) {
-          if (!unique_trxs.insert(trx_h).second) {
-            continue;
-          }
-          db_query.append(DB::Columns::final_chain_transaction_location_by_hash, trx_h);
-          db_query.append(DB::Columns::transactions, trx_h);
-        }
+      auto period_raw = db_->getPeriodDataRaw(period);
+      RLP period_rlp(period_raw);
+      std::vector<Vote> votes;
+      std::vector<DagBlock> dag_blocks;
+      std::vector<Transaction> period_transactions;
+      db_->parsePeriodData(period_rlp, votes, dag_blocks, period_transactions);
+      DB::MultiGetQuery db_query(db_, period_transactions.size());
```

Comment:
as you fixed it, you can remove `transaction_count_hint_` variable

Reply:
I fix this but also there is a lot to refactor/optimize in final chain related to these latest changes. I still don't understand this final chain code that well so I left that as a second step in another PR.